SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialEventTracerUserInterface.cpp

```diff
@@ -0,0 +1,175 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "Interop/Connection/SpatialEventTracerUserInterface.h"
+
+#include "EngineClasses/SpatialNetDriver.h"
+#include "EngineClasses/SpatialPackageMapClient.h"
+#include "Interop/Connection/SpatialEventTracer.h"
+#include "Interop/Connection/SpatialWorkerConnection.h"
+#include "Interop/SpatialClassInfoManager.h"
+#include "SpatialView/EntityComponentId.h"
+
+DEFINE_LOG_CATEGORY(LogSpatialEventTracerUserInterface);
+
+FString USpatialEventTracerUserInterface::CreateSpanId(UObject* WorldContextObject)
+{
+	SpatialGDK::SpatialEventTracer* EventTracer = GetEventTracer(WorldContextObject);
+	if (EventTracer == nullptr || !EventTracer->IsEnabled())
+	{
+		return {};
+	}
+
+	return SpatialGDK::SpatialEventTracer::SpanIdToString(EventTracer->CreateSpan().GetValue());
+}
+
+FString USpatialEventTracerUserInterface::CreateSpanIdWithCauses(UObject* WorldContextObject, const TArray<FString>& Causes)
+{
+	SpatialGDK::SpatialEventTracer* EventTracer = GetEventTracer(WorldContextObject);
+	if (EventTracer == nullptr || !EventTracer->IsEnabled())
+	{
+		return {};
+	}
+
+	TArray<Trace_SpanId> SpanIdS;
+	for (const FString& String : Causes)
+	{
+		SpanIdS.Add(SpatialGDK::SpatialEventTracer::StringToSpanId(String));
+	}
+
+	return SpatialGDK::SpatialEventTracer::SpanIdToString(EventTracer->CreateSpan(SpanIdS.GetData(), SpanIdS.Num()).GetValue());
+}
+
+void USpatialEventTracerUserInterface::TraceEvent(UObject* WorldContextObject, const FString& SpanId, FSpatialTraceEvent SpatialTraceEvent)
+{
+	SpatialGDK::SpatialEventTracer* EventTracer = GetEventTracer(WorldContextObject);
+	if (EventTracer == nullptr)
+	{
+		return;
+	}
+
+	EventTracer->TraceEvent(SpatialTraceEvent, SpatialGDK::SpatialEventTracer::StringToSpanId(SpanId));
+}
+
+void USpatialEventTracerUserInterface::SetActiveSpanId(UObject* WorldContextObject, FEventTracerDynamicDelegate Delegate,
+													   const FString& SpanId)
+{
+	SpatialGDK::SpatialEventTracer* EventTracer = GetEventTracer(WorldContextObject);
+	if (EventTracer == nullptr || !EventTracer->IsEnabled())
+	{
+		return;
+	}
+
+	EventTracer->SpanIdStack.AddNewLayer(SpatialGDK::SpatialEventTracer::StringToSpanId(SpanId));
+	Delegate.Execute();
+	EventTracer->SpanIdStack.PopLayer();
+}
+
+void USpatialEventTracerUserInterface::AddSpanIdToStack(UObject* WorldContextObject, const FString& SpanId)
```

Comment:
I feel like this is too advanced for a user API, typically the stack is used for tracing events through the GDK whereas game code should be able to pass things more directly - unless I missed something

Reply:
Maybe a TraceRPC function could just trace the next called RPC?

Reply:
So this is just used for code user event, not blueprints. Blueprints can user the SetActiveSpanId function. I could add something similar for code, a function that takes like a lambda and wraps up the add and pop stack calls.